### ISSUE TITLE
port tower-governor to rama rate limiting

### ISSUE THREAD

**Issue by GlenDC** [plabayo/rama#88]
Cfr: <https://github.com/benwis/tower-governor>

**Comment by GlenDC** [comment]
Added the "needs input" label. Will add documentation regarding that label in the README in next 24h.
Basically what it means is that it first still needs some R&D and alignment. To save you and I time and frustration.

That work can be done purely by communicating here without any code required (for now).

Questions that needs to be answered in detail and with links:

1. What is `tower-governer` exactly?
2. What is the difference with how we do rate limiting?
3. Are they different or can we perhaps just replace our current rate limiting with that?
4. What dependencies does tower-governer require?

Based on that we can either proceed or cancel. Or think further.

The actual porting, should we get to that point, will be relatively straightforward I think, as tower code is very similar to rama code. What I do usually do when doing such porting is making sure to improve where i can. E.g. within rama where we have async trait support there is no need to handwrite all those futures. So a big simplification that is usually easy to do is to replace hand rolled futures with async fn ones. Just to say some stuff.

But anyway, too early for code for now we need knowledge and alignment. If that's ok with you, by all means do start picking it up @lxrst  and welcome aboard. I'll already assign it to you

**Comment by GlenDC** [comment]
Hmm ok. Need to think about it. If the same to you I would like to shelve this for a while, until there's actual request for this from me or someone else. 

Especially as I usually put rama-based services behind an internal load balancer already that handles this kind of stuff for me on a <=L4 level... So for the time being not something I ever need. Is also the reason why personally I never saw the benefit of having the `poll_ready` stuff in the `Service` interface.

What's your take on this?

**Comment by GlenDC** [comment]
Sounds good.

btw, as mentioned in various places, Rama is modular by design. So rama not supporting better limiting is never a blocker. It’s middleware you can definitely add in a separate repository at any time.

In fact, I think for bigger features like this it might anyway be more easier for contributors to add this in a separate repository first. So that it can already be used for your own projects while you iterate on it and while we figure out if and how we want to integrate it into rama.

so if you do want to still look at it, do feel free to do so, in that way. If you have concerns or more to say about it, do share here, by all means.

**Comment by NOOMA-42** [comment]
@GlenDC Is this still open?

**Comment by GlenDC** [comment]
Yes. It's not an urgent one. Probably easier than the ICAP one as you can start from the tower-governor code, it's MIT licensed. Can you first work out a small plan of attack. Don't need to be technically detailed but more how you see it as user experience, once it would be a rama layer. E.g. small usage example.

And also answer the question: is it possible to define it as a `Policy` implementation for the current Limit layer (https://docs.rs/rama/latest/rama/layer/struct.Limit.html) or do you see it as a replacement of it?

**Comment by NOOMA-42** [comment]
I assume we can't use `GovernorLayer` in `tower_governor` if we're going to reuse `Limit_layer`? With this assumption, I imagine the the user experience is like this: `RateLimitPolicy` will wrap `GovernorConfig`, exposing a `Policy`-compatible interface, so that we ensure the `LimitLayer` remains modular. 
```rust
use tower_governor::governor::GovernorConfigBuilder;
use rama::layer::limit::{LimitLayer, Policy}

impl Policy for RateLimitPolicy {}

let policy = RateLimitPolicy {
    config: GovernorConfigBuilder(
       ...
    ),
};

let layer = LimitLayer::new(policy);
```
However, the pulls-in of several unnecessary modules still remains to be questions to me. Might you shed some light?

**Comment by GlenDC** [comment]
```
use rama::layer::LimitLayer;

// `governor` crate will become a dependency of
// a first community crate that you are free to own
// and publish under your own github handle
//
// I'll list your crate in the book and README.
use rama_x_governor::GovernorPolicy;

// - ideally you make building such a policy infallible
//   by ensuring at compile time that only ever a valid policy can be built (using marker generics where needed)
// - ideally you do not even have to expose governor dep as a public dependency
// - ideally you can hide details away such as spawning a GC background thread,
//   you can do so by doing that only once (first time the policy is used for example)
// - the policy should be cloneable
LimitLayer::new(GovernorPolicy::builder()
    .per_second(2)
    .burst_size(5)
    .gc_interval(Duration::from_secs(60))
    .build());

// You will not depend upon tower_governor, but you can look at it for inspiration
// and copy code from it. You will however depend upon governor.
//
// As it brings however a set of dependencies with it that I do not think are desired
// rama-core, I think it better to be a separate community crate.
```

^ Does that make sense to you?

**Comment by GlenDC** [comment]
Btw for community crates like this I am also available to mentor and guide. So do feel free to post links to the code here and ask for guidance or feedback :) I recommend to also add at least one example in your repo so people know how to use it.

Best of luck.

**Comment by NOOMA-42** [comment]
Just to clarify why do you prefer this as a separate repo instead of a separate folder under rama project ?

**Comment by GlenDC** [comment]
Because it will be a community project, not really something I am for the time being interested in having as a official crate. Because it's partially experimental and partly because I or any of my partners have also no need for it now.

Doesn't mean that it cannot become official in future. But at least this way it can already exist and people can already make use of it without compromising my own choices.

**Comment by NOOMA-42** [comment]
make sense 😄 working on it

**Comment by NOOMA-42** [comment]
@GlenDC I have an initial implementation here https://github.com/NOOMA-42/rama-x-governor/pull/1 , I'd love to know how you think about it 😄

**Comment by GlenDC** [comment]
Looks great so far, keep it up, do let me know once you have it finished or want more feedback.

**Comment by GlenDC** [comment]
As this is moved into a community crate I'm closng this already. Do let me know once it's finished @nooma-42 and works. I'll gladly promote it in the rama docs.

In the meanwhile there's also a crate to provide adapters for tower so that you can use tower services and layers within a rama stack if you need to.
